### PR TITLE
Non markable is on buffered upload

### PR DESCRIPTION
**Pull request by rickle-msft**
This PR allows customers to pass a non-markable InputStream to BlobClient.upload when it would previously throw an UnsupportedOperationException. This method uses buffered upload, so it does not rely on the marking/resetting of the stream the way simple uploads do. Furthermore, it is meant to be a general purpose upload, and v8 also supports nonmarkable streams in this case.

### CODE REVIEW

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java [hunk 1]
```diff
@@ -294,7 +294,7 @@
             final Map<String, String> metadataFinal = options.getMetadata() == null
                 ? new HashMap<>() : options.getMetadata();
             Flux<ByteBuffer> data = options.getDataFlux() == null ? Utility.convertStreamToByteBuffer(
-                options.getDataStream(), options.getLength(), BLOB_DEFAULT_UPLOAD_BLOCK_SIZE)
+                options.getDataStream(), options.getLength(), BLOB_DEFAULT_UPLOAD_BLOCK_SIZE, true)
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java:297]: This should be false

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileAsyncClient.java [hunk 2]
```diff
@@ -318,7 +318,7 @@ public String getFileName() {
             Flux<ByteBuffer> data = options.getDataFlux() == null ? Utility.convertStreamToByteBuffer(
                 options.getDataStream(), options.getLength(),
                 // We can only buffer up to max int due to restrictions in ByteBuffer.
-                (int) Math.min(Integer.MAX_VALUE, validatedParallelTransferOptions.getBlockSizeLong()))
+                (int) Math.min(Integer.MAX_VALUE, validatedParallelTransferOptions.getBlockSizeLong()), true)
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileAsyncClient.java:321]: Should be false

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileClient.java [hunk 3]
```diff
@@ -341,7 +341,7 @@ public void append(InputStream data, long fileOffset, long length) {
 
         Objects.requireNonNull(data);
         Flux<ByteBuffer> fbb = Utility.convertStreamToByteBuffer(data, length,
-            BlobAsyncClient.BLOB_DEFAULT_UPLOAD_BLOCK_SIZE);
+            BlobAsyncClient.BLOB_DEFAULT_UPLOAD_BLOCK_SIZE, true);
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileClient.java:344]: Should add a test for this

File: sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileClient.java [hunk 4]
```diff
@@ -971,7 +971,7 @@ public ShareFileUploadInfo upload(InputStream data, long length) {
     public Response<ShareFileUploadInfo> uploadWithResponse(InputStream data, long length, Long offset,
         ShareRequestConditions requestConditions, Duration timeout, Context context) {
         return StorageImplUtils.blockWithOptionalTimeout(shareFileAsyncClient.uploadWithResponse(Utility
-                .convertStreamToByteBuffer(data, length, (int) ShareFileAsyncClient.FILE_DEFAULT_BLOCK_SIZE),
+                .convertStreamToByteBuffer(data, length, (int) ShareFileAsyncClient.FILE_DEFAULT_BLOCK_SIZE, true),
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileClient.java:974]: Should add a test for this

File: sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/specialized/BlockBlobAPITest.groovy [hunk 5]
```diff
@@ -1433,10 +1471,43 @@ class BlockBlobAPITest extends APISpec {
...
         where:
         dataSizeList                         | blockCount
+        [10, 100, 1000, 10000]               | 0
         [4 * Constants.MB + 1, 10]           | 2
         [4 * Constants.MB, 4 * Constants.MB] | 2
     }
 
+    @Unroll
+    @Requires({ liveMode() })
+    def "Buffered upload sync handle pathing with transient failure"() {
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/specialized/BlockBlobAPITest.groovy:1481]: Can you also make sure the extra length tests get in datalake too?

**Review by gapra-msft — APPROVED** [review]

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/Utility.java [hunk 6]
```diff
@@ -220,22 +220,30 @@ public static OffsetDateTime parseDate(String dateString) {
...
+     * @param markAndReset Whether the stream needs to be marked and reset. This should generally always be true to
+     * support retries. It is false in the case of buffered upload to support non markable streams because buffered
+     * upload uses its own mechanisms to support retries.
      * @return {@link ByteBuffer} which contains the input data.
      * @throws UnexpectedLengthException when input data length mismatch input length.
      * @throws RuntimeException When I/O error occurs.
      */
-    public static Flux<ByteBuffer> convertStreamToByteBuffer(InputStream data, long length, int blockSize) {
-        data.mark(Integer.MAX_VALUE);
+    public static Flux<ByteBuffer> convertStreamToByteBuffer(InputStream data, long length, int blockSize,
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/Utility.java:230]: Given this is a public API should make this a new overload and have the previous method pass the default case which appears to be `true`.

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**